### PR TITLE
chore: add pre-commit config and run linters

### PR DESCRIPTION
**Pull request by lmmx**
> **Background:** Follow-up to #58, which was cancelled as it combined the contribution of pre-commit config (to run linters) with a refactor (for code structure), and it transpired that only the former was desirable upon submission

I've re-done the changes (by copying over the pre-commit exactly as in the previous PR #58 and executing it on this fresh branch). The only edits that were required to achieve a successful linting were the removal of `*` imports (`torchvision.transforms.*` and `uform.models.*` specifically)

Other than that nothing was changed, then the linters ran.

Some of these are the built-in ones from the pre-commit tool's own hooks:

- check-toml
- check-yaml
- debug-statements (to avoid committing breakpoints etc)
- end-of-file-fixer (standardise newlines at end of file)
- name-tests-test (not needed yet as there are no pytest tests but you may wish to include them in future so no harm?)
- trailing-whitespace

Additionally:

- toml-sort-fix (lints `pyproject.toml`)
- ruff (replaces many other tools which are selected with single letter rule category codes in `pyproject.toml`: pyflakes, pycodestyle, isort, pyupgrade)

To run this on CI all you have to do is go to [pre-commit.ci](https://pre-commit.ci/) and give its GitHub Actions app permission to run on your CI (does not require a `.github/workflows` YAML config to run). The pre-commit config YAML file introduced in this PR is sufficient to configure the CI workflow to then run it.

### CODE REVIEW

File: .pre-commit-config.yaml
Comment (ashvardanian) [inline .pre-commit-config.yaml:1]: Would it be a good idea to move this to the `.github` folder to the rest of GitHub-specific configs?

File: LICENSE [hunk 1]
```diff
@@ -198,4 +198,4 @@
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
-   limitations under the License.
\ No newline at end of file
+   limitations under the License.
```
Comment (ashvardanian) [inline LICENSE:201]: Is it possible to exclude this file from linting? I'm afraid some of the GitHub infra may consider this a custom license even if we add as much as a newline.

Comment (lmmx) [reply LICENSE:201]: 1) It is possible to exclude from linting
2) You needn't be afraid of that, this edit was introduced by the "end of file fixer" which I also use in other projects (e.g. [fugit](https://github.com/lmmx/fugit/blob/master/LICENSE.txt) which do not invalidate upon linting.

Licenses aren't detected in a fragile way (like a MD5 checksum/exact match), we can tell that from e.g. MIT licensed projects where the license is detected despite the user editing it (you have to put your name in it for it to be valid).

Specifically the docs on [how licenses are detected on GitHub](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/licensing-a-repository#detecting-a-license) using the open source Ruby gem [licensee](https://github.com/licensee/licensee) using several "matchers" which notes that its method includes

> If we strip away whitespace and copyright notice,

Which confirms that the whitespace change is not going to obscure the license [it does in fact do more than that using string matching but as I say that wouldn't become relevant]

File: .pre-commit-config.yaml
Comment (lmmx) [reply .pre-commit-config.yaml:1]: The `.pre-commit-config.yaml` is not GitHub specific, it's `git` specific, and should live at the project root

> ### [Adding pre-commit plugins to your project ¶](https://pre-commit.com/index.html#adding-pre-commit-plugins-to-your-project)
> 
> Add a file called `.pre-commit-config.yaml` to the root of your project. The pre-commit config file describes what repositories and hooks are installed.

**Pull request merged**